### PR TITLE
Remove node-pre-gyp, use prebuildify

### LINKED ISSUE #665: Move to prebuild
`node-pre-gyp` has been causing a lot of installation errors. Move to [prebuilt](https://github.com/prebuild/prebuild). This should fix most of the installation errors.

_This task is high priority_

### PR DESCRIPTION
**Pull request by thom-nic**
Added Dockerfiles for cross-platform build on linux-x64, arm64 and arm32. ~~Note that on ARM platforms unit tests presently fail :(~~ Tests pass on all architectures (x64,arm,arm64/glibc and musl) and nodejs versions (10,12,14,16 x64/glibc)

Note if you want to run/debug unit tests in an interactive environment it's possible to `docker run -it blah/bcryptjs-linux-armhf-builder` which will start a bash session inside the container.

~~**Edit**: I should mention, the tests have probably always failed on ARM in a way that has nothing to do with prebuildify or docker/ARM emulation.  I did try to checkout the master branch on an arm device and run the tests, and IIRC they failed the same way.  But the good news is, now it's easy to do full cross-platform build *and* test in CI using these arm32 and arm64 docker images.~~ This is all fixed.

**Edit 2**: worth noting, the package/ release process needs to be updated a bit. Unlike node-pre-gyp which would download a platform specific prebuilt archive, prebuildify publishes all prebuilt binaries to NPM (or github nodejs registry).  So while CI could paralellize builds for different platforms, each prebuild artifact needs to be collected into a single unified folder before publishing to NPM. i.e. when you do an `npm publish` your folder structure should look like this:

```
[project root]
 |
 + prebuilds/
    + darwin-x64/
    + linux-arm64/
    + linux-arm/
    + linux-x64/
```
~~So I'm not yet clear on what an automated CI publish process would look like.  Basically you need a sort of "reducer" step that runs after all of the parallel platform-specific prebuilds have completed.  I've been playing with github actions and think there's a workable solution there.~~ See [comment below](#issuecomment-994130606) I have github actions working now.

Fixes #858
Fixes #665

### CODE REVIEW

**Comment by thom-nic** [comment]
Hey bcryptjs maintainers!

I've spent more time on this recently and believe I have a workable solution with github actions (I have it working with node-sqlite3 ported to prebuildify)  It will parallelize tests on multiple nodejs versions and also parallelizes the native build for different platforms (presently linux-{x64,arm,arm64}.  May be possible to build for Alpine/musl too.  So this now has potential to resolve #858 

That said, I still get inconsistent test results between platforms. At the moment, this branch has test failures on Ubuntu 20.04 on x86 but the build passes on my macbook.  🤷

**Comment by thom-nic** [comment]
Ok these appear to be in good shape now.  I added nodejs 10 and alpine builders to the build matrix as well.  You can see the actions run against a PR on my fork here: https://github.com/thom-nic/node.bcrypt.js/pull/1/checks

**Comment by cokia** [comment]
I think this is an important issue too.

Due to the spread of M1 Macs, many developers have begun to use ARM, which is an increasingly high demand problem.

**Comment by recrsn** [comment]
Let's merge it.

Can you rebase it again with some changes in main?

**Comment by recrsn** [comment]
The latest release got rid of nodeunit and tap

**Comment by mrinc** [comment]
@recrsn @thom-nic  just bumping this up

**Comment by thom-nic** [comment]
I will rebase this week

**Comment by mpauly-exnaton** [comment]
@thom-nic I made an attempt at rebasing [over here](https://github.com/mpauly-exnaton/node.bcrypt.js/tree/prebuildify-and-docker-build). 
Things to watch out for:
* I removed github actions as introduced by @recrsn and kept the one introduced by @thom-nic 
* In line with upstream, I dropped support for old node versions and added Node 18 to the build matrix
* I discarded the changes you made to the tests as the tests are now promise-based. I increased the timeouts for the tests. Without this increase the tests were failing.

With these changes the build is passing for me, as you can see [here](https://github.com/mpauly-exnaton/node.bcrypt.js/pull/1). 
I hope this helps to identify the pain-points in rebasing. If you want I can also open a PR with these changes against `master`.

I should also notice that I am by no means an expert in building native node extensions, and just followed your changes. So please have a careful look/feel free to discard everything I did.

**Comment by JasonMan34** [comment]
What exactly is the state of this PR? @thom-nic @recrsn 
If it's a problem of rebasing, is @mpauly-exnaton 's fork any help?

I wouldn't mind forking myself and applying the changes from this PR, or do anything else I can to help promote #665

File: .github/workflows/test.yml
Comment (recrsn) [inline .github/workflows/test.yml:1]: We don't need this, `ci.yaml` already does this for us

File: Dockerfile-alpine
Comment (recrsn) [inline Dockerfile-alpine:1]: Do we need these docker files?

The current CI simply uses Github actions inside docker

**Review by recrsn — CHANGES_REQUESTED** [review]

File: .github/workflows/test.yml
Comment (thom-nic) [reply .github/workflows/test.yml]: Removed.

File: Dockerfile-alpine
Comment (thom-nic) [reply Dockerfile-alpine:1]: They are useful if you want to do xplatform test/ build locally without relying on a GH workflow.  

Otherwise you can't test or build for those targets unless you create a PR and wait for the actions to complete.  It's noticeably slower than running the Docker builds locally as well.

Comment (thom-nic) [reply Dockerfile-alpine:1]: Oh actually they are also used by the build-pack-publish github workflow.  The existing ci workflow does a test on libc and alpine but not for arm/ arm64.

Looking at it now, *maybe* it is possible to do a matrix of jobs with `container.image` like `[ 'arm64v8/node:14-bullseye', 'arm32v7/node:14-bullseye', 'node:14-bullseye']` and also one for alpine, then collect all of the build artifacts for both in the 'pack' step of that workflow.

Comment (thom-nic) [reply Dockerfile-alpine:1]: I guess you can also remove 'test-alpine.sh' since that is not using prebuildify and is covered in the docker build.

**Comment by thom-nic** [comment]
Apparently node-gyp *just* published v10 over the weekend and dropped support for nodejs 14. I pinned the installed version in the dockerfiles.  

FWIW the nodejs version used when building is independent of the runtime nodejs version support.  I verified this by building running the build on node 18 then changing to node v14 and running the test suite.

**Comment by recrsn** [comment]
@thom-nic Thats the advantage of napi, ABI stability. Regardless, we build in the earliest supported nodejs version to prevent any lurking bugs, if any

File: .github/workflows/ci.yaml
Comment (recrsn) [inline .github/workflows/ci.yaml:23]: we are missing `npm ci` here

File: Dockerfile-alpine
Comment (recrsn) [reply Dockerfile-alpine:1]: My concern is mostly around duplicating the Dockerfile logic across all combinations of architecture we plan to support.

One thing we can do here is to have 2 Dockerfiles (one for Debian, one for Alpine) and pass the actual container image as a Dockefile `ARG`

My current way of building Alpine ARM builds is running `test-alpine.sh` under an arm64 docker container.

**Comment by thom-nic** [comment]
@recrsn do you plan to drop the appveyor and travis CI?

File: .github/workflows/ci.yaml
Comment (thom-nic) [reply .github/workflows/ci.yaml:30]: Do you want the docker builders to use `node:18` container image then?

File: Dockerfile-alpine
Comment (thom-nic) [reply Dockerfile-alpine:1]: I think you are referring to [this](https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact) I've never tried that before, let me give it a try.

File: .github/workflows/ci.yaml
Comment (recrsn) [reply .github/workflows/ci.yaml]: Yes

**Comment by thom-nic** [comment]
Ok only using two dockerfiles with `--platform` to build for x86-64 + arm32 + arm64

**Comment by recrsn** [comment]
> @recrsn do you plan to drop the appveyor and travis CI?

Yes

**Comment by thom-nic** [comment]
I think I mostly have CI running correctly.  I have the workflows running on another branch in a private repo and I believe it's working or very close to.  For some reason the arm64 workflow is taking forever at the moment (~20 minutes+) before it fails because the repetition tests timeout.  It was much faster before so I'm not sure what changed or if there's some hiccup in github's infra at the moment.

Worth pointing out, the original `ci.yml` only does tests across node versions but does not do xplatform tests (but those are run as part of the build-pack-publish script.)

**Comment by recrsn** [comment]
Github is releasing arm64 runners soon, so if the tests are taking forever, we can hold it off for a while. 

Meanwhile, I already use a Mac and two RPIs to test releases.

**Comment by thom-nic** [comment]
Ok I have the build and test running on all platforms, without hardcoding an exceptionally long timeout - only during the workflow.  Given that it all passes now, it would be really nice to merge this *now* and if arm64 runners can be utilized in the future, change it then. 

I don't really want to have to address this issue again in the future.   

Screenshot 2023-11-02 at 9 35 40 AM

**Review by recrsn — APPROVED** [review]
This looks good now

**Comment by AaronMoat** [comment]
@recrsn would it be possible to cut a release for this, please? It happens to also remove a vulnerable dependency, which I'm sure will start to nag people's CI systems :) 

https://security.snyk.io/vuln/SNYK-JS-INFLIGHT-6095116

**Pull request merged**
